### PR TITLE
Speeding up adding factors + computing wiring

### PR DESCRIPTION
**Pull request by antoine-dedieu**
Main goal of this PR is to speed up the process of adding factors to a FactorGraph and to compile wiring. In particular
- we do not instantiate factors in a FactorGraph by default. Instead all the computations, including wiring, happen at the factor groups level. 
- all the factors-related operations are computed on demand, when the user requires them
- we create a subfolder `pgmax/groups/` to clarify the structure

Before:
- adding factors for the RBM exp takes 25s, building run_bp takes 45s
- adding factors for the convor exp takes 20s, building run_bp takes 25s

After:
- adding factors for the RBM exp takes 2s, building run_bp takes 4s
- adding factors for the convor exp takes 2s, building run_bp takes 10s

Our next step will accelerate the wiring compilation further by  using `numba`

### CODE REVIEW

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:37]: Might be better to make this an attribute of the corresponding factor group classes instead of having to maintain a centralized registry, something like
```
factor_type: type = field(init=False, default=enumeration.EnumerationFactor)
```

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:154]: Why this change? This function is meant to add single factors imperatively, so there should only be one set of `variable_names`

Comment (StannisZhou) [reply pgmax/fg/graph.py:154]: I see it's to make all factor groups have the same consistent interface. That is fine, but to avoid confusion, the argument for `add_factor_by_type` should still be just `variable_names`. Inside the function we can wrap it inside a list and pass the list to `SingleFactorGroup`.

After the change `SingleFactorGroup` should have an additional check to make sure its `len(variable_names_for_factors) == 1`.

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:566]: If you do this you might want to add `num_factors: int = field(init=False)` to the class definition. Otherwise mypy will complain when you re-enable the check

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:584]: Same as above. Add to class definition

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:600]: This is a property so you shouldn't have the `()` in the end

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:629]: Why is there a full version? If this is the non-flat version it should just be used to define the class, and shouldn't be a property

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:695]: Do you mean `== 1`?

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:792]: Isn't this exactly the same as `self.log_potentials`?

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:839]: Can we keep all the relevant logic inside this function, instead of defining a function elsewhere?

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:1026]: Again not sure this property is necessary

File: pgmax/fg/graph.py
Comment (antoine-dedieu) [reply pgmax/fg/graph.py:154]: totally agree

File: pgmax/fg/nodes.py
Comment (wlehrach) [inline pgmax/fg/nodes.py:1]: I wouldn't drop these (actually, I'd always add them) - there are all sorts of annoyances that can crop up with ordering of modules/functions inside files and type annotations.

Comment (StannisZhou) [reply pgmax/fg/nodes.py:1]: If mypy is not complaining then probably fine to get rid of some of these? It's only useful for some of the modules and with mypy check enforced we don't need to worry about missing failure cases

File: pgmax/groups/logical.py
Comment (StannisZhou) [inline pgmax/groups/logical.py:110]: Since `factor_type` is an attribute, it looks like this function can be implemented just one within the `LogicalFactorGroup`, with factor creations done by `self.factor_type(...)`?

File: pgmax/groups/logical.py
Comment (StannisZhou) [inline pgmax/groups/logical.py:63]: This seems really the same as the previous concatenation approach. The only difference is whether we create the individual factor objects. Curious to know what the performance difference is....

Also, since we are just manipulating arrays, it seems easy to make this whole function parallel using numba. Can be done in the next PR!

File: pgmax/groups/enumeration.py
Comment (StannisZhou) [inline pgmax/groups/enumeration.py:28]: I'm guessing just `factor_configs: np.ndarray` is fine (since we are overriding parent class)?

File: pgmax/groups/enumeration.py
Comment (StannisZhou) [inline pgmax/groups/enumeration.py:29]: Get rid of `field`. Also why do we need to change the original `Optional[np.ndarray] = None` here? `None` makes more sense than empty in this case

File: pgmax/groups/enumeration.py
Comment (StannisZhou) [inline pgmax/groups/enumeration.py:94]: This compiling function is really the same as the `compile_wiring` function in individual enumeration factors. There should not be two different implementations of the same function. Since we are trying to consolidate, can we define a function in `factors.enumeration` that handles the wiring compilation for all 3 cases?

Comment (StannisZhou) [reply pgmax/groups/enumeration.py:94]: In addition, it still feels wrong to have to repeat this function multiple times.

One thought for possibly defining this function only once (inside the parent class): we still take the compile then concatenate approach, but we create a functional implementation for wiring compilation for individual factors (which can be used without instantiating individual factors). We then wrap things in numba to do the individual compilation in parallel before the concatenation.

The same approach should also work for LogicalFactors (and in factor any new factors we add) without change.

We can discuss more and try to tackle this in the next PR. For this PR fine to just consolidate all 3 places into one function.

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:372]: Add `default=np.empty((0,))` and get rid of the `if` in `__post_init__`.

File: pgmax/fg/groups.py
Comment (StannisZhou) [inline pgmax/fg/groups.py:369]: It seems it's better to change these to `cached_property`s (e.g. `factor_edges_num_states` to be consistent with how it's done for individual factors, and `variables_for_factors` which is the parallel of `variables` in individual factors.

File: pgmax/groups/enumeration.py
Comment (StannisZhou) [inline pgmax/groups/enumeration.py:25]: log_potentials should also be documented (move the documentation currently in the parent class to here)

File: pgmax/groups/enumeration.py
Comment (StannisZhou) [inline pgmax/groups/enumeration.py:203]: If we get rid of `factor_configs` in the parent class we don't need to do this awkward thing here

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:79]: Move this back to avoid unnecessary diffs

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:104]: Remove before merging

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:109]: Why change `None` to empty?

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:203]: We probably don't want to remove. I added this check after accidentially adding the same factor twice and getting really confused about the results. I think it's a useful check....

File: pgmax/fg/graph.py
Comment (StannisZhou) [inline pgmax/fg/graph.py:217]: Nit: this is really just factor_type. Factor groups don't have types....

**Review by StannisZhou — COMMENTED** [review]
Made some detailed comments, although nothing major.

It looks like there should be a way to use numba parallel and significantly speed up wiring compilation while unifying the compile wiring interface for all factor groups. So although the current factor group wiring compilation is still a bit annoying, we can go with what we have now after some additional cleanups (since we are likely going to rewrite many of these in the next PR).

File: pgmax/fg/groups.py
Comment (antoine-dedieu) [reply pgmax/fg/groups.py:371]: This is for mypy.
`factor_configs` has to be specified for EnumerationFactor, so if we only keep the log_potentials in the parent Factor class, we will get `Non default argument cannot follow a default one`

File: pgmax/fg/groups.py
Comment (antoine-dedieu) [reply pgmax/fg/groups.py:369]: I would advocate to not moving to `cached_property` to only loop through the factors once, and save time

File: pgmax/fg/nodes.py
Comment (antoine-dedieu) [reply pgmax/fg/nodes.py:89]: yes: both `logical.py` and `enumeration.py` have `concatenate_wirings` before `compile_wiring` so we should make this consistent in the parent class

File: pgmax/groups/enumeration.py
Comment (antoine-dedieu) [reply pgmax/groups/enumeration.py:29]: the parent class will assing log_potentials to `np.empty((0,))` by default, so adding the None option is adding noise

File: pgmax/groups/enumeration.py
Comment (antoine-dedieu) [reply pgmax/groups/enumeration.py:203]: Same, it's mypy related

File: pgmax/groups/logical.py
Comment (antoine-dedieu) [reply pgmax/groups/logical.py:63]: This is much faster as we do not instantiate the factors objects.
Totally agree, the next PR will speed this up using numba!

File: tests/fg/test_graph.py
Comment (antoine-dedieu) [reply tests/fg/test_graph.py:38]: this means we need to change `_register_factor_group`? let's discuss this

File: pgmax/fg/graph.py
Comment (antoine-dedieu) [reply pgmax/fg/graph.py:109]: As I mentioned, for more consistency

File: pgmax/groups/enumeration.py
Comment (antoine-dedieu) [reply pgmax/groups/enumeration.py:96]: I agree that this function should only be called once.
I am planning to move this to numba in the next PR, let's keep it as it is for now

File: pgmax/fg/graph.py
Comment (antoine-dedieu) [reply pgmax/fg/graph.py:79]: The comment `# See FactorGraphState docstrings for documentation on the following fields` does not apply to this object

**Comment by antoine-dedieu** [comment]
Thanks for your thorough review @StannisZhou 
There are a few points we should discuss related to mypy behaviour
I totally agree that the next PR should be focused on speeding the factor group wiring using numba!

File: tests/fg/test_graph.py
Comment (StannisZhou) [reply tests/fg/test_graph.py:38]: I updated `fg.graph` for this part. Please update this test accordingly!

The fix right now is not entirely correct for LogicalFactors, but a proper fix seems unnecessarily complicated. I will create a issue to keep track of this but for now we can go with the not entirely correct solution using frozenset

**Review by StannisZhou — APPROVED** [review]
Left a few more minor comments. LGTM after addressing the comments and making all tests pass!

**Pull request merged**
